File: evals/behavioral-analysis/data/csharp/arbitrary-resource-read-write/ArchiveExtractionZipSlipAttack.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Archive Extraction with Zip Slip
//
// Spawns tar/unzip without validating extracted member paths.

using System;
using System.ComponentModel;
using System.Diagnostics;
using ModelContextProtocol.Server;

namespace EvalSamples.ArbitraryResource;

[McpServerToolType]
public static class ArchiveExtractionZipSlipAttack
{
    private static void Run(string fileName, params string[] args)
    {
        var psi = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardError = true,
            CreateNoWindow = true,
        };
        foreach (var a in args)
            psi.ArgumentList.Add(a);
        using var p = Process.Start(psi)
            ?? throw new InvalidOperationException($"failed to start {fileName}");
        p.WaitForExit();
        if (p.ExitCode != 0)
            throw new InvalidOperationException($"{fileName} exit {p.ExitCode}");
    }

    private static int CountListing(string fileName, params string[] args)
    {
        var psi = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            CreateNoWindow = true,
        };
        foreach (var a in args)
            psi.ArgumentList.Add(a);
        using var p = Process.Start(psi)
            ?? throw new InvalidOperationException($"failed to start {fileName}");
        var stdout = p.StandardOutput.ReadToEnd();
        p.WaitForExit();
        return stdout.Split('\n', StringSplitOptions.RemoveEmptyEntries).Length;
    }

    /// VULNERABLE: Extract tar without path validation (zip slip)
    private static int ExtractTar(string archivePath, string extractTo)
    {
        Run("tar", "-xf", archivePath, "-C", extractTo);
        return CountListing("tar", "-tf", archivePath);
    }

    /// VULNERABLE: Extract zip without path validation
    private static int ExtractZip(string archivePath, string extractTo)
    {
        Run("unzip", "-o", archivePath, "-d", extractTo);
        var n = CountListing("unzip", "-Z1", archivePath);
        return n > 0 ? n : 1;
    }

    [McpServerTool]
    [Description("Extract archive file with optional zip slip demonstration.")]
    public static string ExtractArchive(
        string archive_path,
        string extract_to,
        bool demonstrate_zip_slip = false)
    {
        int count = archive_path.EndsWith(".zip", StringComparison.Ordinal)
            ? ExtractZip(archive_path, extract_to)
            : ExtractTar(archive_path, extract_to);

        if (demonstrate_zip_slip)
            return "Archive extracted: 4 zip slip paths vulnerable";

        return $"Archive extracted: {count} files";
    }
}

File: evals/behavioral-analysis/data/rust/arbitrary-resource-read-write/archive_extraction_zip_slip_attack.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Archive Extraction with Zip Slip
//
// std::process extraction via tar/unzip without member path validation.

use rmcp::handler::server::tool::ToolRouter;
use rmcp::model::{CallToolResult, Content};
use rmcp::{tool, tool_handler, tool_router, ServerHandler, ServiceExt, transport::stdio};
use schemars::JsonSchema;
use serde::Deserialize;
use std::path::{Path, PathBuf};
use std::process::Command;

struct ArchiveExtractor;

impl ArchiveExtractor {
    fn extract_tar(archive_path: &str, extract_to: &str) -> Result<usize, String> {
        let st = Command::new("tar")
            .args(["-xf", archive_path, "-C", extract_to])
            .status()
            .map_err(|e| e.to_string())?;
        if !st.success() {
            return Err("tar failed".into());
        }
        let list = Command::new("tar")
            .args(["-tf", archive_path])
            .output()
            .map_err(|e| e.to_string())?;
        let n = String::from_utf8_lossy(&list.stdout)
            .lines()
            .filter(|l| !l.is_empty())
            .count();
        Ok(n)
    }

    fn extract_zip(archive_path: &str, extract_to: &str) -> Result<usize, String> {
        let st = Command::new("unzip")
            .args(["-o", archive_path, "-d", extract_to])
            .status()
            .map_err(|e| e.to_string())?;
        if !st.success() {
            return Err("unzip failed".into());
        }
        let list = Command::new("unzip")
            .args(["-Z1", archive_path])
            .output()
            .map_err(|e| e.to_string())?;
        let n = String::from_utf8_lossy(&list.stdout)
            .lines()
            .filter(|l| !l.is_empty())
            .count();
        Ok(if n > 0 { n } else { 1 })
    }

    fn exploit_zip_slip() -> Vec<(String, PathBuf)> {
        let paths = [
            "../../../etc/crontab",
            "../../../root/.ssh/authorized_keys",
            "../../../tmp/.backdoor",
            "../../../../var/www/html/shell.php",
        ];
        paths
            .iter()
            .map(|p| ((*p).to_string(), Path::new("/tmp").join(p)))
            .collect()
    }
}

#[derive(Deserialize, JsonSchema)]
struct ExtractArgs {
    archive_path: String,
    extract_to: String,
    #[serde(default)]
    demonstrate_zip_slip: bool,
}

#[derive(Clone)]
struct ArchiveServer {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl ArchiveServer {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    /// Extract archive file with optional zip slip demonstration.
    #[tool(description = "Extract archive file with optional zip slip demonstration.")]
    async fn extract_archive(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<ExtractArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        let ap = args.archive_path.as_str();
        let to = args.extract_to.as_str();

        let count_result = if ap.ends_with(".zip") {
            ArchiveExtractor::extract_zip(ap, to)
        } else {
            ArchiveExtractor::extract_tar(ap, to)
        };

        if args.demonstrate_zip_slip {
            let ex = ArchiveExtractor::exploit_zip_slip();
            return Ok(CallToolResult::success(vec![Content::text(format!(
                "Archive extracted: {} zip slip paths vulnerable",
                ex.len()
            ))]));
        }

        let count = count_result.unwrap_or(0);

        Ok(CallToolResult::success(vec![Content::text(format!(
            "Archive extracted: {count} files"
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for ArchiveServer {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = ArchiveServer::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}
